tf2onnx/onnx_opset/tensor.py

```diff
@@ -2364,32 +2364,45 @@ def version_10(cls, ctx, node, **kwargs):
 
 
 @tf_op("Unique", onnx_op="Unique")
+@tf_op("UniqueWithCounts", onnx_op="Unique")
 class Unique:
+    int_cast = [TensorProto.BOOL, TensorProto.INT32, TensorProto.INT16, TensorProto.UINT8,
+                TensorProto.UINT16, TensorProto.UINT32, TensorProto.UINT64]
+    dtype_map = {k: TensorProto.INT64 for k in int_cast}
+    dtype_map[TensorProto.DOUBLE] = TensorProto.FLOAT
+
     @classmethod
     def version_11(cls, ctx, node, **kwargs):
         # opset 11 supports explicitly
-        dtypes = node.output_dtypes
         node_name = node.name
         node_inputs = node.input
         node_outputs = node.output
+        inp_dtype = ctx.get_dtype(node.input[0])
+
         ctx.remove_node(node_name)
-        if dtypes[0] in [TensorProto.INT32, TensorProto.INT16, TensorProto.UINT8, TensorProto.UINT16]:
-            inp_cast = ctx.make_node("Cast", [node_inputs[0]], attr={'to': TensorProto.INT64}).output[0]
+
+        # due to ORT missing implementations we need to cast INT inputs to INT64 and FLOAT inputs to FLOAT32
+        if inp_dtype in cls.dtype_map:
+            inp_cast = ctx.make_node("Cast", [node_inputs[0]], attr={'to': cls.dtype_map[inp_dtype]}).output[0]
             node_inputs[0] = inp_cast
-        new_node = ctx.make_node("Unique", node_inputs, name=node_name, output_count=3, attr={'sorted': 0})
+
+        new_node = ctx.make_node("Unique", node_inputs, name=node_name, attr={'sorted': 0},
+                                 outputs=[utils.make_name("y"), utils.make_name("idx_first"),
+                                          utils.make_name("idx"), utils.make_name("counts")])
         ctx.replace_all_inputs(node_outputs[0], new_node.output[0])
         ctx.replace_all_inputs(node_outputs[1], new_node.output[2])
-        if ctx.get_dtype(new_node.output[0]) != dtypes[0]:
-            ctx.insert_new_node_on_output("Cast", new_node.output[0], name=utils.make_name(node.name) + "_cast",
-                                          to=dtypes[0])
-        if len(node_outputs) > 1:
-            # cast to int64 if needed
-            if dtypes[1] != onnx_pb.TensorProto.INT64:
-                cast_node = ctx.insert_new_node_on_output("Cast", new_node.output[2],
-                                                          name=utils.make_name(node.name) + "_cast",
-                                                          to=dtypes[1])
-                ctx.set_dtype(cast_node.output[0], dtypes[1])
-                ctx.copy_shape(new_node.output[2], cast_node.output[0])
+        if len(node_outputs) == 3: # we need counts too (UniqueWithCounts)
+            ctx.replace_all_inputs(node_outputs[2], new_node.output[3])
+        if ctx.get_dtype(new_node.output[0]) != inp_dtype:
+            ctx.insert_new_node_on_output("Cast", new_node.output[0], to=inp_dtype,
+                                          name=utils.make_name(node.name) + "_cast")
+
+        # cast idx and counts if needed
```

Comment:
Where do we cast 'counts'?

Reply:
In the the for loop (Line 2403). If the node has 2 outputs (TF "Unique" op) we cast onnx  output 2 ("inverse_indices"). If the node has 3 outputs (TF "UniqueWithCounts" op) we cast onnx output 2 ("inverse_indices") and output 3 ("counts"). Either case, we skip onnx output 1 ("indices") since it is not present  in the TF ops.